File: week4_tutorial/IPC/message passing/go/tcp_server.go
package main

import (
	"fmt"
	"net"
)

func main() {
	// Replace with the actual IP address of the server
	serverIP := "server_ip_here"
	serverPort := 12345

	// Listen for incoming TCP connections on the specified IP and port
	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", serverIP, serverPort))
	if err != nil {
		fmt.Println("Error listening:", err)
		return
	}
	defer listener.Close() // Close the listener when the function exits

	fmt.Printf("Server listening on %s:%d\n", serverIP, serverPort)

	// Accept a new incoming connection
	conn, err := listener.Accept()
	if err != nil {
		fmt.Println("Error accepting connection:", err)
		return
	}
	defer conn.Close() // Close the connection when the function exits

	// Create a buffer to hold received data
	buffer := make([]byte, 1024)

	// Read data from the connection into the buffer
	n, err := conn.Read(buffer)
	if err != nil {
		fmt.Println("Error reading:", err)
		return
	}

	// Print the received data as a string
	fmt.Println("Received:", string(buffer[:n]))
}

File: week4_tutorial/IPC/message passing/rust/tcp_server.rs
use std::io::{Read, Write};
use std::net::{TcpListener, TcpStream};

fn main() {
    // Replace with the actual IP address of the server
    let server_ip = "server_ip_here";
    let server_port = 12345;

    // Bind a TCP listener to the specified IP and port
    let listener = TcpListener::bind(format!("{}:{}", server_ip, server_port)).unwrap();
    println!("Server listening on {}:{}", server_ip, server_port);

    // Start accepting incoming connections in a loop
    for stream in listener.incoming() {
        match stream {
            Ok(mut stream) => {
                // Create a buffer to read incoming data
                let mut buffer = [0; 1024];
                match stream.read(&mut buffer) {
                    Ok(n) => {
                        // Convert received bytes to a string and print
                        let received_data = String::from_utf8_lossy(&buffer[..n]);
                        println!("Received: {}", received_data);
                    }
                    Err(e) => {
                        // Error occurred while reading data from the stream
                        println!("Error reading: {:?}", e);
                    }
                }
            }
            Err(e) => {
                // Error occurred while accepting a connection
                println!("Error accepting connection: {:?}", e);
            }
        }
    }
}
